### ISSUE TITLE
solveset() doesn't support symbolic functions yet

### ISSUE THREAD

**Issue by roger-** [sympy/sympy#12239]
Example with SymPy 1.0 with Python 3.6:

```python
from sympy import solveset 
import sympy as sm

x  = sm.Symbol('x')  
f  = sm.Function('f')  

# this works 
sm.solve(f(0)**2 - 1, f(0))

# so does this
sm.nsolve([f(0)**2 - 1], [f(0)], [3])

# this doesn't
solveset(f(1)**2 - 1, f(1))

# but this does
solveset(x**2 - 1, x)
```
The second to last example gives:

```python
NotImplementedError: 
relationship between value and 0 is unknown: Eq(f(1)**2 - 1, 0)
```

So I guess this hasn't been added yet. 

This might be related to another inconsistency in symbolic functions in #12169.

**Comment by grassknoted** [comment]
I would like on this issue. I would be really helpful is @Upabjojr or @asmeurer could tell me how to get started.

**Comment by Upabjojr** [comment]
@grassknoted in such cases, I usually start by writing some simple test cases and then follow the execution of the code with the debugger. Then you'd try to guess from the point the exception was raised.

**Comment by Yathartha22** [comment]
I would like to take this issue.

**Comment by Yathartha22** [comment]
@Upabjojr I was thinking to implement it in a manner that `solve` does. Is this a good idea?
If so could you tell me the start point(could be the line in the code) in the `solvers` module where `solve` checks for this particular condition

**Comment by grassknoted** [comment]
@Yathartha22  I am working on this issue, if you suggest your fix, we could work together.

**Comment by Yathartha22** [comment]
I was thinking that if we can replace the `function` term (`f(0)`) with a `dummy` variable then probably we could solve this like normal equations. I don't know if this is right or wrong but this was the first thing that came to my mind.
what do you say?

**Comment by Upabjojr** [comment]
> I was thinking that if we can replace the function term (f(0)) with a dummy variable then probably we could solve this like normal equations.

That's a quick way to fix it. It's probably due to the fact that functions do not show up as free symbols.

**Comment by Yathartha22** [comment]
Should I start working on this idea then @Upabjojr

**Comment by Upabjojr** [comment]
You can propose any PR you feel like.

**Comment by grassknoted** [comment]
That's a novel approach, but wouldn't that reduce the flexibility of `solveset()`?

**Comment by smichr** [comment]
`solveset` already will solve for symbolic functions but a re-ordering of code is necessary to handle the case when `symbol` is not a Symbol is needed, however, especially in the case when solving for a function having no symbols in the arguments, e.g. `solveset(f(1) - 3, f(1))`.

The code can be updated to use `recast_to_symbols`

There is no need to impose additional constraints on what non-symbol object can be solved for.

The following diff can be applied to master and it...
 
fixes #12239
closes #12245
closes #13811


```diff
diff --git a/sympy/solvers/solveset.py b/sympy/solvers/solveset.py
index 59b60b2..5e12af5 100644
--- a/sympy/solvers/solveset.py
+++ b/sympy/solvers/solveset.py
@@ -898,28 +898,26 @@ def solveset(f, symbol=None, domain=S.Complexes):
 
     free_symbols = f.free_symbols
 
-    if not free_symbols:
-        b = Eq(f, 0)
-        if b is S.true:
-            return domain
-        elif b is S.false:
-            return S.EmptySet
-        else:
-            raise NotImplementedError(filldedent('''
-                relationship between value and 0 is unknown: %s''' % b))
-
     if symbol is None:
         if len(free_symbols) == 1:
             symbol = free_symbols.pop()
-        else:
+        elif free_symbols:
             raise ValueError(filldedent('''
                 The independent variable must be specified for a
                 multivariate equation.'''))
     elif not isinstance(symbol, Symbol):
-        s = Dummy()
-        f = f.subs(symbol, s)
+        f, s, swap = recast_to_symbols([f], [symbol])
         # the xreplace will be needed if a ConditionSet is returned
-        return solveset(f, s, domain).xreplace({s: symbol})
+        return solveset(f[0], s[0], domain).xreplace(swap)
+    elif not free_symbols:
+        b = Eq(f, 0)
+        if b is S.true:
+            return domain
+        elif b is S.false:
+            return S.EmptySet
+        else:
+            raise NotImplementedError(filldedent('''
+                relationship between value and 0 is unknown: %s''' % b))
 
     if isinstance(f, Eq):
         from sympy.core import Add
diff --git a/sympy/solvers/tests/test_solveset.py b/sympy/solvers/tests/test_solveset.py
index a4527e6..b2f2628 100644
--- a/sympy/solvers/tests/test_solveset.py
+++ b/sympy/solvers/tests/test_solveset.py
@@ -893,9 +893,15 @@ def test_solve_lambert():
 
 def test_solveset():
     x = Symbol('x')
+    f = Function('f')
     raises(ValueError, lambda: solveset(x + y))
     assert solveset(x, 1) == S.EmptySet
+    assert solveset(f(1)**2 + y + 1, f(1)
+        ) == FiniteSet(-sqrt(-y - 1), sqrt(-y - 1))
+    assert solveset(f(1)**2 - 1, f(1), S.Reals) == FiniteSet(-1, 1)
+    assert solveset(f(1)**2 + 1, f(1)) == FiniteSet(-I, I)
     assert solveset(x - 1, 1) == FiniteSet(x)
+    assert solveset(sin(x) - cos(x), sin(x)) == FiniteSet(cos(x))
 
     assert solveset(0, domain=S.Reals) == S.Reals
     assert solveset(1) == S.EmptySet
```

**Comment by ArighnaIITG** [comment]
Okay.
@smichr , I will be working on this issue.
